### ISSUE TITLE
`multi-line-candidate` is a void variable in HEAD of `multi-line` package

### ISSUE THREAD

**Issue by bryce-carson** [syl20bnr/spacemacs#15255]
#### Description :octocat:
`multi-line-candidate` is a void variable in `multi-line-cycle.el` after IvanMalison/multi-line@7c5fbaea5216949820ba8a8d5969d87f36d7d41d.

#### Reproduction guide :beetle:
- Checkout syl20bnr/spacemacs@d63e172bde5b2b3cb8ddb676f6db1e6df4b31fbb, or the @HEAD commit of the develop branch
- Start Emacs with or without `--debug-init`

*Observed behaviour:* :eyes: :broken_heart:

- After initialization, without `--debug-init`, the active buffer will change to the *Warnings* buffer and the void-variable error is displayed.
- With `--debug-init` the backtrace is displayed.

> `Error (use-package): multi-line/:catch: Symbol’s value as variable is void: multi-line-candidate`

*Expected behaviour:* :heart: :smile:
Emacs initializes without errors related to the recent commit to the [`multi-line`](https://github.com/IvanMalison/multi-line/commit/7c5fbaea5216949820ba8a8d5969d87f36d7d41d) package.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 27.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. d63e172bd)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(html auto-completion better-defaults emacs-lisp git helm lsp org python syntax-checking systemd themes-megapack treemacs yaml javascript
      (ess :variables ess-r-backend 'lsp ess-use-company 't)
      (shell :variables shell-default-height 30 shell-default-position 'bottom)
      (sql :variables sql-backend 'lsp sql-lsp-sqls-workspace-config-path nil))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER GMP

#### Backtrace :paw_prints:
[GitHub Gist of the backtrace](https://gist.github.com/bryce-carson/3b98231cf0a3ea79f5bee1802dcccbb3)

**Comment by lebensterben** [comment]
@bryce-carson 
Should this be reported upstream?

**Comment by bryce-carson** [comment]
@lebensterben Yes, we should report this upstream. I'll make an issue upstream and refer to this one.

For now, is it possible for the develop branch to freeze the multi-line package to IvanMalison/multi-line@d5ae863ced0adeb7032ada398005f27a6c669d79? Is this a troubleshooting step that Spacemacs users will need to take themselves?

**Comment by lebensterben** [comment]
@bryce-carson 
Our package declaration is very similar to quelpa's syntax. So I think it's possible to use a specific GitHub repo+commit to pinning down it.

**Comment by bryce-carson** [comment]
I tested the package with a vanilla Emacs and found no issue installing or using the package as per the instructions of `multi-line` after enabling Melpa.

1. `emacs -Q`
2. Enable MELPA
3. Install `multi-line`
4. "C-c d" with point within some ELisp in a scratch buffer.

**Comment by lebensterben** [comment]
any idea why it complained about the void variable?

**Comment by bryce-carson** [comment]
I would say it's related to the change from `eieio` to `cl-lib` in more files, and the - [ ] change from `defmethod` to `cl-defmethod`.

I'm not sure why that would cause problems in Spacemacs, but it's been a forever-bug that `cl` is deprecated. Perhaps we need to investigate the trace of usages of `cl-defmethod` in Spacemacs to find the issue.

**Comment by lebensterben** [comment]
@bryce-carson 
recently we had a commit to migrate to cl-defmethod/cl-defgeneric for our own code.

So I don't think that's the cause.

**Comment by bryce-carson** [comment]
Looking at the head commit of `multi-line`, the only difference is that cl-defmethod is used.

Perhaps the previous commits are culprit.

**Comment by bryce-carson** [comment]
> @bryce-carson 
> recently we had a commit to migrate to cl-defmethod/cl-defgeneric for our own code.
> 
> So I don't think that's the cause.

Okay, that's quite strange then. Perhaps there's something else going on.

**Comment by prehensilecode** [comment]
Am seeing the same using:

### System Info 💻
- OS: macOS 10.15.7
- Emacs: 27.2
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. f60aec5)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers: 
```elisp
(lua (python :variables python-backend 'anaconda) vimscript helm 
 emacs-lisp common-lisp git github markdown org python 
 ipython-notebook (shell :variables shell-default-height 30 
  shell-default-position 'bottom shell-default-shell 'eshell) 
 ranger version-control 
 (org :variables org-roam-v2-ack t org-enable-roam-support t 
  org-enable-org-brain-support t org-enable-notifications t 
  org-todo-dependencies-strategy `naive-auto)
)
```

**Comment by lebensterben** [comment]
Failed to reproduce the error on any of the last few commit of Spacemacs.
(And I've also updated all my packages.)

**Comment by bryce-carson** [comment]
I still get the same error after updating Spacemacs and my packages.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 27.2
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. c3ad21645)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(html auto-completion better-defaults emacs-lisp git helm lsp org python syntax-checking systemd themes-megapack treemacs yaml javascript
      (ess :variables ess-r-backend 'lsp ess-use-company 't)
      (shell :variables shell-default-height 30 shell-default-position 'bottom)
      (sql :variables sql-backend 'lsp sql-lsp-sqls-workspace-config-path nil))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER GMP

**Comment by lebensterben** [comment]
From the number of upvote OP get, I can say this definitely happened to some people.
But I'm not able to reproduce it.

Help needed.

**Comment by bryce-carson** [comment]
> @bryce-carson Our package declaration is very similar to quelpa's syntax. So I think it's possible to use a specific GitHub repo+commit to pin it down.

I followed [Syl's instructions](https://emacs.stackexchange.com/a/20512) for using a local or recipe version of a package rather than what is included in a layer (ie `+spacemacs/spacemacs-editing`).

1. Using the recipe below and freezing the package to IvanMalison/multi-line@d5ae863ced0adeb7032ada398005f27a6c669d79, before the package was updated to fix problems with Emacs 29 (mostly just by changing to `cl-defmethod` and dropping `eieieo`) gets rid of the warning on Develop (0.999.0 @ c3ad216).

```elisp
(multi-line :location (recipe
                       :fetcher github
                       :repo "IvanMalison/multi-line"
                       :commit "d5ae863ced0adeb7032ada398005f27a6c669d79"))
```
Add the above to the `dotspacemacs-additional-packages '()` list and unsure to add `(use-package multi-line)` to the `dotspacemacs/user-config` defun.

2. In a shell of your choice, remove the cached and ELPA packages to ensure that Spacemacs will fetch the package and rebuild it if you're using native-comp or anything.
```sh
find ~/.emacs.d -type d -name "multi-line*" -exec rm -Rf {} +
```

This doesn't solve the issue with upstream, but it does make it more compelling that the issue lies there unless there is some disparate user-config atom or variable somewhere in multiple Spacemacs users builds that are causing `multi-line` to fail in this way.

**Comment by karlredman** [comment]
Disclaimer: I haven't touched emacs/elisp in 15 years and hope this isn't a red herring

tl;dr (details below):
* I'm (fresh) installing spacemacs for the first time ever via MX Linux Debian backports emacs version 27.1.
*  After getting a warning for  `evil-want-keybinding' was set to nil but not before loading evil` I restarted emacs and the warning went away.
* On restart I got the issue related to CL being depreciated, I forget the ticket number.
* I added the fix `(setq byte-compile-warnings '(cl-functions))` at the top of my `init.el` file and restarted emacs again.
* Upon emacs restart I get the `Error (use-package): multi-line/:catch: Symbol’s value as variable is void: multi-line-candidate` error.
* The most recent reference I find for `emacs.d` (dotfiles) for `multi-line` is 2021-02-17 by @tris.
* Anecdotally, this leads me to believe, relative to the postes before me, that this issue is upstream -in agreement with @bryce-carson.

System Info:
* OS: MX Linux 19.2
* Emacs: 27.1
* Spacemacs:0.999.0
* Editing Style: vim
* Completion: helm
* Spacemacs branch: develop (3ad21645)
* Graphics display: t and g
* Editing Style: vim
* Completion: helm

`strings $(command -v emacs) | grep Id`:
$Id: GNU Emacs 27.1 (x86_64-pc-linux-gnu XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP) $


`system-configuration-optoins`:
"--build x86_64-linux-gnu --prefix=/usr --sharedstatedir=/var/lib --libexecdir=/usr/lib --localstatedir=/var/lib --infodir=/usr/share/info --mandir=/usr/share/man --enable-libsystemd --with-pop=yes --enable-locallisppath=/etc/emacs:/usr/local/share/emacs/27.1/site-lisp:/usr/local/share/emacs/site-lisp:/usr/share/emacs/27.1/site-lisp:/usr/share/emacs/site-lisp --with-sound=alsa --without-gconf --with-mailutils --build x86_64-linux-gnu --prefix=/usr --sharedstatedir=/var/lib --libexecdir=/usr/lib --localstatedir=/var/lib --infodir=/usr/share/info --mandir=/usr/share/man --enable-libsystemd --with-pop=yes --enable-locallisppath=/etc/emacs:/usr/local/share/emacs/27.1/site-lisp:/usr/local/share/emacs/site-lisp:/usr/share/emacs/27.1/site-lisp:/usr/share/emacs/site-lisp --with-sound=alsa --without-gconf --with-mailutils --with-cairo --with-x=yes --with-x-toolkit=gtk3 --with-toolkit-scroll-bars 'CFLAGS=-g -O2 -fdebug-prefix-map=/emacs/emacs=. -fstack-protector-strong -Wformat -Werror=format-security -Wall' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' LDFLAGS=-Wl,-z,relro"

**Comment by theophilusx** [comment]
Just to add, I'm also seeing the same issue with a completely fresh clone of spacemacs and all default settings. On restarting Emacs I get the error mentioned, Default configuration, accepting defaults for both the init questions and no other changes. 

This is on Ubuntu 21.10
Emacs: 28.0.91 built today
Spacemacs current HEAD develop branch

**Comment by afwlehmann** [comment]
As per @bryce-carson 's above suggestion I tried the following and it solved the issue for me (develop/`c3ad21645`):

> ```shell
> find ~/.emacs.d -type d -name "multi-line*" -exec rm -Rf {} +
> ```

**Comment by bryce-carson** [comment]
> As per @bryce-carson 's above suggestion I tried the following and it solved the issue for me (develop/`c3ad21645`):
> 
> > ```shell
> > find ~/.emacs.d -type d -name "multi-line*" -exec rm -Rf {} +
> > ```

Did you also freeze the package to the earlier version as per step 1, or did you only need to delete the package in the cache and the ELPA folder?

**Comment by Khoulaiz** [comment]
I have the same issue. Just deleting the multi-line from the melpa cache did not help. I also needed to freeze the package as described to get rid of the problem.

**Comment by bryce-carson** [comment]
> I have the same issue. Just deleting the multi-line from the melpa cache did not help. I also needed to freeze the package as described to get rid of the problem.

Did you use the `find` command or remove it manually? I had multiple folders 📂 of `multi-line` in my Emacs dot–folder, but I suspect that was because I tried installing a different version in a couple ways before following Syl's correct method.

**Comment by Khoulaiz** [comment]
I used the find command. But I did the freezing wrongly in my first attempt and got the same issue with the one and only freshly downloaded melpa folder. I deleted this one again, corrected the freezing statement (forgot the use-package entry) and then it worked with the pinned version.

**Comment by bryce-carson** [comment]
> @bryce-carson 
> Our package declaration is very similar to quelpa's syntax. So I think it's possible to use a specific GitHub repo+commit to pin it down.

It seems enough users are encountering this that we might consider a PR to conditionally use the earlier version of `multi-line` if the Emacs version is not version 29, delete the cache if we need to upon that condition being true, and otherwise allow the HEAD to be used.

**Comment by lebensterben** [comment]
@bryce-carson 
Go ahead and make a PR.

**Comment by thanhvg** [comment]
It's the byte-compile problem, it should be fixed on multi-line side. Their latest change breaks byte compilation. They should keep (require 'eioio) because they use defclass.  (require 'cl-lib) won't bring deflcass in.

**Comment by bryce-carson** [comment]
> It's the byte-compile problem, it should be fixed on multi-line side. Their latest change breaks byte compilation. They should keep (require 'eioio) because they use defclass. (require 'cl-lib) won't bring defclass in.

Excellent. Could you communicate that to upstream in this reopened issue? IvanMalison/multi-line#18

**Comment by prehensilecode** [comment]
Latest commits have fixed this for me (macOS 10.15.7, Emacs 27.2, spacemacs 0.999.0 develop c3ad216). Thanks!

**Comment by bryce-carson** [comment]
> Latest commits have fixed this for me (macOS 10.15.7, Emacs 27.2, spacemacs 0.999.0 develop [c3ad216](https://github.com/syl20bnr/spacemacs/commit/c3ad21645c7e48fec50f6ca29f4ecd831f8023f5)). Thanks!

@lebensterben, this is interesting. Do you know of any reason why recent commits might fix the issue?

**Comment by lebensterben** [comment]
@bryce-carson 
I'm watching this threads and I'm completely lost.

**Comment by prehensilecode** [comment]
> Latest commits have fixed this for me (macOS 10.15.7, Emacs 27.2, spacemacs 0.999.0 develop [c3ad216](https://github.com/syl20bnr/spacemacs/commit/c3ad21645c7e48fec50f6ca29f4ecd831f8023f5)). Thanks!

I think I have a guess as to what happened, and it may not be latest commits that fixed it. It's only a guess because I was not tracking closely what I was doing, but it's a reasonable guess. Based on shell history, I ran `find ~/.emacs.d -type d -name "multi-line*" -exec rm -rf {} \;` at some point. So, it must have been right before I launched Emacs and saw that the error no longer occured.

Apologies for the confusion.

**Comment by afwlehmann** [comment]
> > As per @bryce-carson 's above suggestion I tried the following and it solved the issue for me (develop/`c3ad21645`):
> > > ```shell
> > > find ~/.emacs.d -type d -name "multi-line*" -exec rm -Rf {} +
> > > ```
> 
> Did you also freeze the package to the earlier version as per step 1, or did you only need to delete the package in the cache and the ELPA folder?

Sorry for the late reply. I did not do the freeze, only the removal of packages from the cache using `find`.

**Comment by manokara** [comment]
Can confirm that removing the cache with `find` solved the error for me as well.

**Comment by thanhvg** [comment]
Upstream has been updated with a fix, please  update the package again on Spacemacs.

**Comment by pataquets** [comment]
Confirming that updating package to latest version from Spacemacs' `SPC f e U` fixes issue on Ubuntu.
